### PR TITLE
Fixes Bug With Fire and Acid Not Properly Updating Snow Stack Visuals

### PR DESCRIPTION
**Pull request by Surrealaser**
## About The Pull Request

Fire and acid now both properly update snow stack visuals.

## Why It's Good For The Game

Fixes a bug.

Fixes: https://github.com/tgstation/TerraGov-Marine-Corps/issues/5766

## Changelog
:cl:
fix: Fire and acid now both properly update snow stack visuals.
refactor: Refactored elements of acidspray_act and related procs.
/:cl:

### CODE REVIEW

File: code/game/atoms.dm [hunk 1]
```diff
@@ -333,6 +333,10 @@ directive is properly returned.
 /atom/proc/fire_act()
 	return
 
+
+/atom/proc/acidspray_act()
+
+
```
Comment (TiviPlus) [inline code/game/atoms.dm:339]: This proc should not exist on atom

File: code/game/objects/items/bodybag.dm [hunk 2]
```diff
@@ -232,16 +232,14 @@
 			visible_message("<span class='danger'>\The shockwave blows [name] apart!</span>")
 			qdel(src) //blown apart
 
-/obj/structure/closet/bodybag/proc/acidspray_act()
-	SIGNAL_HANDLER
+/obj/structure/closet/bodybag/acidspray_act(datum/source, obj/effect/xenomorph/spray/acid_puddle)
 	if(!opened && bodybag_occupant)
-		var/obj/effect/xenomorph/spray/S = locate() in range(0, src) //get the acid Hans
-		if(!S) //Sanity
+		if(!source) //Sanity
```
Comment (TiviPlus) [inline code/game/objects/items/bodybag.dm:237]: no.

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/__DEFINES/dcs/signals.dm [hunk 3]
```diff
@@ -127,7 +127,7 @@
 #define COMSIG_ATOM_INITIALIZED_ON "atom_initialized_on"		//called from atom/Initialize() of target: (atom/target)
 #define COMSIG_ATOM_ORBIT_BEGIN "atom_orbit_begin"				//called when an atom starts orbiting another atom: (atom)
 #define COMSIG_ATOM_ORBIT_STOP "atom_orbit_stop"				//called when an atom stops orbiting another atom: (atom)
-#define COMSIG_ATOM_ACIDSPRAY_ACT "atom_acidspray_act"			//from base of atom/acidspray_act():
+#define COMSIG_ATOM_ACIDSPRAY_ACT "atom_acidspray_act"			//from base of atom/acidspray_act()
```
Comment (TiviPlus) [inline code/__DEFINES/dcs/signals.dm:130]: Also add missing args here instead of removing :

File: code/game/turfs/snow.dm [hunk 4]
```diff
@@ -168,19 +163,22 @@
...
+			update_icon(1, 0)
 		if(11 to 24)
 			slayer = max(0, slayer - 2)
+			update_icon(1, 0)
 		if(25 to INFINITY)
 			slayer = 0
+			update_icon(1, 0)
 
 /turf/open/floor/plating/ground/snow/proc/acidspray_act()
-	SIGNAL_HANDLER
```
Comment (SplinterGP) [inline code/game/turfs/snow.dm:178]: why remove that signal handler ?

File: code/game/turfs/snow.dm [hunk 5]
```diff
@@ -168,19 +163,22 @@
...
 		if(1 to 10)
-			slayer--
+			slayer = max(0, slayer - 1)
+			update_icon(1, 0)
 		if(11 to 24)
 			slayer = max(0, slayer - 2)
+			update_icon(1, 0)
 		if(25 to INFINITY)
 			slayer = 0
+			update_icon(1, 0)
```
Comment (SplinterGP) [inline code/game/turfs/snow.dm:173]: why repeat the update_icon everytime ? just put it out of the chain

**Review by SplinterGP — CHANGES_REQUESTED** [review]

File: code/game/objects/effects/aliens.dm [hunk 6]
```diff
@@ -94,10 +94,11 @@
 		qdel(src)
 		return
 
-	for(var/mob/living/carbon/human/H in loc)
+	for(var/mob/living/carbon/human/H in T)
 		H.acid_spray_crossed(acid_damage, slow_amt)
```
Comment (SplinterGP) [inline code/game/objects/effects/aliens.dm:98]: typechecking
use
for(var/H in T)
    var/mob/living/carbon/human/H2 = H

**Review by SplinterGP — CHANGES_REQUESTED** [review]

File: code/game/turfs/snow.dm (hunk 5 above)
Comment (Surrealaser) [reply code/game/turfs/snow.dm:173]: Copy pasta tbh but yeah

Comment (SplinterGP) [reply code/game/turfs/snow.dm:173]: if possible do on the other one that you copypasted from

File: code/game/turfs/snow.dm (hunk 4 above)
Comment (Surrealaser) [reply code/game/turfs/snow.dm:178]: It works, but it's ultimately not necessary and linters yells at me

**Comment by SplinterGP** [comment]
checking about the linters issue, merge soon.

File: code/game/objects/effects/aliens.dm [hunk 7]
```diff
@@ -94,12 +94,16 @@
...
+	SEND_SIGNAL(T, COMSIG_ATOM_ACIDSPRAY_ACT, src) //Signal the turf
+	for(var/H in T)
 
-	for(var/atom/A in loc) //Infrastructure for other interactions
+		var/atom/A = H
 		SEND_SIGNAL(A, COMSIG_ATOM_ACIDSPRAY_ACT, src)
 
+		if(ishuman(A))
+			var/mob/living/carbon/human/H2 = H
+			H2.acid_spray_crossed(acid_damage, slow_amt)
```
Comment (TiviPlus) [inline code/game/objects/effects/aliens.dm:105]: this should use the signal

**Review by TiviPlus — CHANGES_REQUESTED** [review]

**Comment by TiviPlus** [comment]
Also use signal handlers for signal procs

File: code/game/objects/effects/aliens.dm (hunk 7 above)
Comment (Surrealaser) [reply code/game/objects/effects/aliens.dm:105]: This is basically how it was, I just made it less retarded.

**Comment by Surrealaser** [comment]
> Also use signal handlers for signal procs

Bruh, I can't. It works as is, and if I try to add SIGNAL_HANDLER, linters yells at me.

Look at where I told linters to fuck off.

File: code/game/objects/effects/aliens.dm [hunk 8]
```diff
@@ -59,43 +59,45 @@
...
+	SIGNAL_HANDLER
 
-/mob/living/carbon/human/proc/acid_spray_crossed(acid_damage, slow_amt)
 	if(TIMER_COOLDOWN_CHECK(src, COOLDOWN_ACID))
 		return
 
 	TIMER_COOLDOWN_START(src, COOLDOWN_ACID, 1 SECONDS)
 	if(HAS_TRAIT(src, TRAIT_FLOORED))
-		take_overall_damage_armored(acid_damage, BURN, "acid", updating_health = TRUE)
+		INVOKE_ASYNC(src, .proc/take_overall_damage_armored, acid_damage, BURN, "acid", FALSE, FALSE, TRUE)
```
Comment (SplinterGP) [inline code/game/objects/effects/aliens.dm:73]: i believe you shoulnd't need this, i think what's sleeping should be only emotes

Comment (Surrealaser) [reply code/game/objects/effects/aliens.dm:73]: I need it; linters will NOT accept it off async; you can look it up, the chain eventually hits sleep

Comment (SplinterGP) [reply code/game/objects/effects/aliens.dm:73]: i'll kill sleep() whole career

File: code/game/objects/effects/aliens.dm [hunk 9]
```diff
@@ -59,43 +59,45 @@
...
 	var/armor_block = run_armor_check(affecting, "acid")
-	if(istype(affecting) && affecting.take_damage_limb(0, acid_damage/2, FALSE, FALSE, armor_block, TRUE))
-		UpdateDamageIcon()
+	INVOKE_ASYNC(affecting, /datum/limb/.proc/take_damage_limb, 0, acid_damage/2, FALSE, FALSE, armor_block)
+
 	affecting = get_limb(BODY_ZONE_PRECISE_R_FOOT)
 	armor_block = run_armor_check(affecting, "acid")
-	if(istype(affecting) && affecting.take_damage_limb(0, acid_damage/2, FALSE, FALSE, armor_block, TRUE, updating_health = TRUE))
-		UpdateDamageIcon()
+	INVOKE_ASYNC(affecting, /datum/limb/.proc/take_damage_limb, 0, acid_damage/2, FALSE, FALSE, armor_block, TRUE)
```
Comment (SplinterGP) [inline code/game/objects/effects/aliens.dm:87]: what about health updates ?

Comment (Surrealaser) [reply code/game/objects/effects/aliens.dm:87]: Exactly; only need to call it the last time.

**Review by TiviPlus — APPROVED** [review]

**Pull request merged**